### PR TITLE
refactor: make maintenance service controller-based

### LINKED ISSUE #7430: Refactor maintenance service to be controller-driven
1. Have configuration for the maintenance service - when it needs to run, which interfaces to listen on, etc.
2. Have a controller to manage TLS certs for the maintenance service.
3. Run maintenance service using a controller, reconfigure it when maintenance config changes.

### PR DESCRIPTION
**Pull request by smira**
Fixes #7430

Introduce a set of resources which look similar to other API implementations: CA, certs, cert SANs, etc.

Introduce a controller which manages the service based on resource state.

### CODE REVIEW

File: pkg/machinery/config/container/container.go [hunk 1]
```diff
@@ -226,16 +226,22 @@ func (container *Container) Validate(mode validation.RuntimeMode, opt ...validat
 		warnings, err = container.v1alpha1Config.Validate(mode, opt...)
 	}
 
+	var multiErr *multierror.Error
```
Comment (smira) [inline pkg/machinery/config/container/container.go:229]: this is a bug I found with multi-doc config validation (unrelated to this PR), regression tests added

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+		Logger: logger.Sugar().Debugf,
+	}
+	tlsProvider := maintenance.NewTLSProvider()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case cfg := <-cfgCh:
+			configResource := config.NewMachineConfigWithID(cfg, config.MaintenanceID)
```
Comment (Unix4ever) [inline internal/app/machined/pkg/controllers/runtime/maintenance_service.go:122]: Why can't we replace all that with `safe.WriterModify(...)`?

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+				if err = configResource.Metadata().SetOwner(oldConfigResource.Metadata().Owner()); err != nil {
+					return fmt.Errorf("error setting owner: %w", err)
+				}
+
+				if err = r.Update(ctx, configResource); err != nil {
+					return fmt.Errorf("failed to update machine config: %w", err)
+				}
+			}
+
+			continue
```
Comment (Unix4ever) [inline internal/app/machined/pkg/controllers/runtime/maintenance_service.go:145]: Can we replace all that with `safe.WriterModify(...)`?

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+					grpc.Creds(
+						credentials.NewTLS(tlsConfig),
+					),
+				),
+
+				factory.WithUnaryInterceptor(injector.UnaryInterceptor()),
+				factory.WithStreamInterceptor(injector.StreamInterceptor()),
+			)
+
+			go func() {
```
Comment (Unix4ever) [inline internal/app/machined/pkg/controllers/runtime/maintenance_service.go:226]: can we add that to errgoup and do `eg.Wait()` in `shutdownServer` to be on the save side?

**Review by Unix4ever — APPROVED** [review]

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go (hunk 4 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/runtime/maintenance_service.go:226]: I can look into that, need to make sure we can re-use the waitgroup, as the controller might start the server multiple times

File: internal/app/machined/pkg/controllers/secrets/maintenance_cert_sans.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+			spec.AppendIPs(nodeAddresses.TypedSpec().IPs()...)
+			spec.AppendIPs(netip.MustParseAddr("127.0.0.1"))
+			spec.AppendIPs(netip.MustParseAddr("::1"))
+
+			if hostnameStatus != nil {
+				spec.AppendDNSNames(hostnameStatus.TypedSpec().DNSNames()...)
+			}
+
+			spec.FQDN = "maintenance.talos.dev"
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/secrets/maintenance_cert_sans.go:99]: constants maybe? 
```suggestion
			spec.FQDN = "maintenanceservice.talos.dev"
```

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go (hunk 3 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/runtime/maintenance_service.go:145]: this resource is a bit special, it doesn't support changing the spec, and WriterModify requires a "null" value which is not really defined for machine config

**Review by frezbo — APPROVED** [review]

File: internal/app/machined/pkg/controllers/runtime/maintenance_service.go (hunk 4 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/runtime/maintenance_service.go:232]: fixed, added `sync.WaitGroup`

File: internal/app/machined/pkg/controllers/secrets/maintenance_cert_sans.go (hunk 5 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/secrets/maintenance_cert_sans.go]: fixed, thanks!

**Pull request merged**